crypto/cms/cms_sd.c

```diff
@@ -223,10 +224,39 @@ int cms_SignerIdentifier_cert_cmp(CMS_SignerIdentifier *sid, X509 *cert)
         return -1;
 }
 
-static int cms_sd_asn1_ctrl(CMS_SignerInfo *si, int cmd)
+static int cms_sd_asn1_ctrl(CMS_SignerInfo *si, EVP_PKEY_CTX *pctx, int cmd)
 {
     EVP_PKEY *pkey = si->pkey;
+    X509_ALGOR *alg = NULL;
     int i;
+
+    if (pkey->keymgmt == NULL)
+        goto legacy;
+
+    /*
+     * Key algorithms that should lead to an AlgorithmIdentifier with
+     * the combined keyalg-with-digestalg OID should respond to the
+     * evp_pkey_ctx_get_algid().  When the OID should specify the keyalg
+     * only, the response is allowed to come from either that or
+     * evp_pkey_get_algid().
+     */
+    alg = evp_pkey_ctx_get_algid(pctx, OSSL_CMS_PARAM_ALGORITHM_ID);
+    if (alg == NULL)
+        alg = evp_pkey_get_algid(pkey, OSSL_CMS_PARAM_ALGORITHM_ID);
+    /* Try the common X509_ALGOR as a fallback */
+    if (alg == NULL)
+        alg = evp_pkey_ctx_get_algid(pctx, OSSL_PKEY_PARAM_ALGORITHM_ID);
```

Comment:
So we ask the pctx using the cms specific param, then the pkey using the cms specific param, and then the pctx using the generic param. But we never ask the pkey using the generic param. Why do we never ask the pkey for the generic param?

Reply:
I'm wondering if that's needed. How's an RSA key response for CMS (OID rsaEncryption) going to differ a response for X.509 (OID rsaEncryption)? Do we expect to see anything where that will differ?

But on the other, maybe as well cover all bases just to be safe, is that what you're thinking?

Reply:
> Do we expect to see anything where that will differ?

Perhaps not. I'm not sure.

> But on the other, maybe as well cover all bases just to be safe, is that what you're thinking?

Well yes - it was just that it stuck out as being a bit strange. I could live with the answer being that we don't think that is ever needed...as long we have actually thought about it and believe that is most likely to be the case.

Reply:
I can flesh out the comment if that helps

Reply:
ok

Reply:
Actually, I'm doubling back on this.  The more I look at it, the more I see it as an omission.